### PR TITLE
PICARD-2912: Script editor syntax highlighting fixes and improvements

### PR DESCRIPTION
**Pull request by zas**
# Summary

* This is a…
  * [x] Bug fix
  * [x] Feature addition
  * [ ] Refactoring
  * [ ] Minor / simple change (like a typo)
  * [ ] Other
* **Describe this change in 1-2 sentences**:

# Problem

Capture d’écran du 2024-05-24 17-26-14

* JIRA ticket (_optional_): PICARD-2912

# Solution

Capture d’écran du 2024-05-24 19-44-13

# Action

Additional actions required:
* [ ] Update Picard [documentation](https://github.com/metabrainz/picard-docs) (please include a reference to this PR)
* [ ] Other (please specify below)

### CODE REVIEW

**Review by rdswift — DISMISSED** [review]

File: picard/ui/widgets/scripttextedit.py [hunk 1]
```diff
@@ -101,44 +101,66 @@ class TaggerScriptSyntaxHighlighter(QtGui.QSyntaxHighlighter):
...
         self.noop_fmt = QtGui.QTextCharFormat()
         self.noop_fmt.setFontWeight(QtGui.QFont.Weight.Bold)
         self.noop_fmt.setFontItalic(True)
         self.noop_fmt.setForeground(syntax_theme.noop)
-        self.rules = [
-            (self.func_re, self.func_fmt, 0, -1),
+
+        self.rules = list(self.func_rules())
+        self.rules.extend([
+            (self.unknown_func_re, self.unknown_func_fmt, 0, -1),
```
[the remark this answers is not in the document]
Comment (zas) [reply picard/ui/widgets/scripttextedit.py]: @rdswift I went for classes instead of namedtuple finally, check last commits

**Review by rdswift — DISMISSED** [review]
Looks good to me.  Thanks for using a class to clarify the rules items.  That makes it a lot easier to understand.

File: picard/ui/theme.py [hunk 2]
```diff
@@ -87,7 +87,7 @@ def _missing_(cls, value):
     func=QtGui.QColor(255, 87, 160, 255),  # magenta
     var=QtGui.QColor(252, 187, 81, 255),  # orange
     escape=QtGui.QColor(75, 239, 31, 255),  # green
-    special=QtGui.QColor(255, 87, 160, 255),  # blue
+    special=QtGui.QColor(150, 150, 239),  # blue
     noop=QtGui.QColor(4, 231, 213, 255),  # cyan
```
Comment (phw) [inline picard/ui/theme.py:91]: I'm not entirely sure about this change. I find the blueish color hard to read and don't really see why we need to use a separate color for the parentheses and commas. But if we change it, shouldn't we also change it in the light theme? Maybe I also don't fully understand what this intents to fix.

If it is only about highlighting invalid functions I would suggest to introduce a separate "error" color just for this.

Small side note: For consistency I would either remove the alpha from all values or keep it in this value as well.

**Review by phw — CHANGES_REQUESTED** [review]

**Comment by phw** [comment]
Good move with highlighting and refactoring it with the classes. I only don't quite get the color change, especially as it is in dark mod only.

I also added PICARD-2912 to track the change of highlighting missing functions.

File: picard/ui/theme.py (hunk 2 above)
Comment (zas) [reply picard/ui/theme.py:91]: Perhaps you missed the point: it has a different color, but in light mode only. I see no reason to have same color in dark mode, while they differ in light mode.

Comment (zas) [reply picard/ui/theme.py:91]: > If it is only about highlighting invalid functions

It was here before, not for invalid functions only.

> I would suggest to introduce a separate "error" color just for this.

Yes, I thought about that too, but first I'd like to make those colors configurable.

Comment (phw) [reply picard/ui/theme.py:91]: > Perhaps you missed the point: it has a different color, but in light mode only.

Not it hasn't, it's `QtCore.Qt.GlobalColor.blue` for both function names and the parentheses / commas. That's why this change confused me.

Comment (zas) [reply picard/ui/theme.py:91]: Ah damn, I was confused by the comment on https://github.com/metabrainz/picard/blob/e4168b53859ab8f4c2b5fa205f292349770a3d3c/picard/ui/theme.py#L90 (vs https://github.com/metabrainz/picard/blob/e4168b53859ab8f4c2b5fa205f292349770a3d3c/picard/ui/theme.py#L87)

Ok, so we have 2 identical colors, I'll revert this change.

Comment (zas) [reply picard/ui/theme.py]: Done & rebased

Comment (phw) [reply picard/ui/theme.py]: I see. I think that's an example of a comment not following code changes. I likely added the comments first before replacing the colors, and as the light theme uses "blue" here the comment was blue. When changing the actual color to magenta only the comment for the first use was updated. So we should fix the comment here, I did so in f78d02e415d89b608735548d8426e5d69dccb3c9

**Review by phw — APPROVED** [review]
LGTM

**Pull request merged**